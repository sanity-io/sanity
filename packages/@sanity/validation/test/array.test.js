const {Rule} = require('../src')

describe('array', () => {
  test('required constraint', async () => {
    const rule = Rule.array().required()
    await expect(rule.validate(null)).resolves.toMatchSnapshot('required: null')
    await expect(rule.validate(undefined)).resolves.toMatchSnapshot('required: undefined')
    await expect(rule.validate([])).resolves.toMatchSnapshot('required: empty array')
    await expect(rule.validate(['hei'])).resolves.toMatchSnapshot('required: valid')
  })

  test('min length constraint', async () => {
    const rule = Rule.array().min(2)
    await expect(rule.validate(['a'])).resolves.toMatchSnapshot('min length: too short')
    await expect(rule.validate(['a', 'b', 'c'])).resolves.toMatchSnapshot('min length: valid')
  })

  test('max length constraint', async () => {
    const rule = Rule.array().max(2)
    await expect(rule.validate(['a', 'b', 'c', 'd'])).resolves.toMatchSnapshot(
      'max length: too long'
    )
    await expect(rule.validate(['a'])).resolves.toMatchSnapshot('max length: valid')
  })

  test('exact length constraint', async () => {
    const rule = Rule.array().length(2)
    await expect(rule.validate(['a', 'b', 'c'])).resolves.toMatchSnapshot('exact length: too long')
    await expect(rule.validate(['a'])).resolves.toMatchSnapshot('exact length: too short')
    await expect(rule.validate(['a', 'b'])).resolves.toMatchSnapshot('exact length: valid')
  })

  test('unique constraint (default, simple values)', async () => {
    const rule = Rule.array().unique()
    await expect(rule.validate(['a', 'b', 'c', 'd'])).resolves.toMatchSnapshot(
      'simple unique: valid'
    )
    await expect(rule.validate(['a', 'b', 'c', 'a'])).resolves.toMatchSnapshot(
      'simple unique: duplicates'
    )
  })

  test('unique constraint (default, object values)', async () => {
    const rule = Rule.array().unique()
    const ref = (id) => ({_ref: id, _type: 'reference'})
    await expect(rule.validate(['a', 'b', 'c', 'd'].map(ref))).resolves.toMatchSnapshot(
      'object unique: valid'
    )
    await expect(rule.validate(['a', 'b', 'c', 'a'].map(ref))).resolves.toMatchSnapshot(
      'object unique: duplicates'
    )
  })

  test('unique constraint (default, array values)', async () => {
    const rule = Rule.array().unique()
    const refArr = (id) => [{_ref: id, _type: 'reference'}]
    await expect(rule.validate(['a', 'b', 'c', 'd'].map(refArr))).resolves.toMatchSnapshot(
      'array unique: valid'
    )
    await expect(rule.validate(['a', 'a', 'c', 'd'].map(refArr))).resolves.toMatchSnapshot(
      'array unique: duplicates'
    )
  })

  test('unique constraint (default, bool values)', async () => {
    const rule = Rule.array().unique()
    await expect(rule.validate([true, false])).resolves.toMatchSnapshot('boolean unique: valid')
    await expect(rule.validate([false, true, false])).resolves.toMatchSnapshot(
      'boolean unique: duplicates'
    )
  })

  test('unique constraint (default, numeric values)', async () => {
    const rule = Rule.array().unique()
    await expect(rule.validate([1, 3])).resolves.toMatchSnapshot('numeric unique: valid')
    await expect(rule.validate([3, 1, 3])).resolves.toMatchSnapshot('numeric unique: duplicates')
  })
})
