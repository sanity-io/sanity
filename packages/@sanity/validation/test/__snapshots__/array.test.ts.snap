// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`array exact length constraint: exact length: too long 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Must have exactly 2 items",
      "operation": undefined,
      "paths": Array [],
    },
    "level": "error",
    "path": Array [],
    "type": "validation",
  },
]
`;

exports[`array exact length constraint: exact length: too short 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Must have exactly 2 items",
      "operation": undefined,
      "paths": Array [],
    },
    "level": "error",
    "path": Array [],
    "type": "validation",
  },
]
`;

exports[`array exact length constraint: exact length: valid 1`] = `Array []`;

exports[`array max length constraint: max length: too long 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Must have at most 2 items",
      "operation": undefined,
      "paths": Array [],
    },
    "level": "error",
    "path": Array [],
    "type": "validation",
  },
]
`;

exports[`array max length constraint: max length: valid 1`] = `Array []`;

exports[`array min length constraint: min length: too short 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Must have at least 2 items",
      "operation": undefined,
      "paths": Array [],
    },
    "level": "error",
    "path": Array [],
    "type": "validation",
  },
]
`;

exports[`array min length constraint: min length: valid 1`] = `Array []`;

exports[`array required constraint: required: empty array 1`] = `Array []`;

exports[`array required constraint: required: null 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Expected type \\"Array\\", got \\"null\\"",
      "operation": undefined,
      "paths": Array [],
    },
    "level": "error",
    "path": Array [],
    "type": "validation",
  },
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Required",
      "operation": undefined,
      "paths": Array [],
    },
    "level": "error",
    "path": Array [],
    "type": "validation",
  },
]
`;

exports[`array required constraint: required: undefined 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Required",
      "operation": undefined,
      "paths": Array [],
    },
    "level": "error",
    "path": Array [],
    "type": "validation",
  },
]
`;

exports[`array required constraint: required: valid 1`] = `Array []`;

exports[`array unique constraint (default, array values): array unique: duplicates 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Can't be a duplicate",
      "operation": undefined,
      "paths": Array [
        Array [
          0,
        ],
        Array [
          1,
        ],
      ],
    },
    "level": "error",
    "path": Array [
      0,
    ],
    "type": "validation",
  },
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Can't be a duplicate",
      "operation": undefined,
      "paths": Array [
        Array [
          0,
        ],
        Array [
          1,
        ],
      ],
    },
    "level": "error",
    "path": Array [
      1,
    ],
    "type": "validation",
  },
]
`;

exports[`array unique constraint (default, array values): array unique: valid 1`] = `Array []`;

exports[`array unique constraint (default, bool values): boolean unique: duplicates 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Can't be a duplicate",
      "operation": undefined,
      "paths": Array [
        Array [
          0,
        ],
        Array [
          2,
        ],
      ],
    },
    "level": "error",
    "path": Array [
      0,
    ],
    "type": "validation",
  },
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Can't be a duplicate",
      "operation": undefined,
      "paths": Array [
        Array [
          0,
        ],
        Array [
          2,
        ],
      ],
    },
    "level": "error",
    "path": Array [
      2,
    ],
    "type": "validation",
  },
]
`;

exports[`array unique constraint (default, bool values): boolean unique: valid 1`] = `Array []`;

exports[`array unique constraint (default, numeric values): numeric unique: duplicates 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Can't be a duplicate",
      "operation": undefined,
      "paths": Array [
        Array [
          0,
        ],
        Array [
          2,
        ],
      ],
    },
    "level": "error",
    "path": Array [
      0,
    ],
    "type": "validation",
  },
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Can't be a duplicate",
      "operation": undefined,
      "paths": Array [
        Array [
          0,
        ],
        Array [
          2,
        ],
      ],
    },
    "level": "error",
    "path": Array [
      2,
    ],
    "type": "validation",
  },
]
`;

exports[`array unique constraint (default, numeric values): numeric unique: valid 1`] = `Array []`;

exports[`array unique constraint (default, object values): object unique: duplicates 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Can't be a duplicate",
      "operation": undefined,
      "paths": Array [
        Array [
          0,
        ],
        Array [
          3,
        ],
      ],
    },
    "level": "error",
    "path": Array [
      0,
    ],
    "type": "validation",
  },
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Can't be a duplicate",
      "operation": undefined,
      "paths": Array [
        Array [
          0,
        ],
        Array [
          3,
        ],
      ],
    },
    "level": "error",
    "path": Array [
      3,
    ],
    "type": "validation",
  },
]
`;

exports[`array unique constraint (default, object values): object unique: valid 1`] = `Array []`;

exports[`array unique constraint (default, simple values): simple unique: duplicates 1`] = `
Array [
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Can't be a duplicate",
      "operation": undefined,
      "paths": Array [
        Array [
          0,
        ],
        Array [
          3,
        ],
      ],
    },
    "level": "error",
    "path": Array [
      0,
    ],
    "type": "validation",
  },
  Object {
    "item": ValidationError {
      "children": undefined,
      "message": "Can't be a duplicate",
      "operation": undefined,
      "paths": Array [
        Array [
          0,
        ],
        Array [
          3,
        ],
      ],
    },
    "level": "error",
    "path": Array [
      3,
    ],
    "type": "validation",
  },
]
`;

exports[`array unique constraint (default, simple values): simple unique: valid 1`] = `Array []`;
